airbrakes/data_handling/data_processor.py

```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
+        ):
+            # quaternions are too uncertain to use, so we calculate them from gyro
+            current_orientation = self._current_orientation_quaternions
+            # Iterates through the data points and time differences between the data points
+            for idx, (data_packet, dt) in enumerate(
+                zip(self._data_packets, self._time_differences, strict=True)
+            ):
+                # Accelerations are in m/s^2
+                x_accel = compensated_accelerations[idx][0]
+                y_accel = compensated_accelerations[idx][1]
+                z_accel = compensated_accelerations[idx][2]
+                # Angular rates are in rads/s
+                gyro_x = data_packet.estAngularRateX
+                gyro_y = data_packet.estAngularRateY
+                gyro_z = data_packet.estAngularRateZ
+
+                # Check for missing data points
+                if any(val is None for val in [gyro_x, gyro_y, gyro_z]):
+                    return rotated_accelerations
+
+                # scipy docs for more info: https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html
+                # Calculate the delta quaternion from the angular rates
+                delta_rotation = R.from_rotvec([gyro_x * dt, gyro_y * dt, gyro_z * dt])
+
+                # Update the current orientation by applying the delta rotation
+                current_orientation = current_orientation * delta_rotation
+
+                # Rotate the acceleration vector using the updated orientation
+                rotated_accel = current_orientation.apply([x_accel, y_accel, z_accel])
+                # Vertical acceleration will always be the 3rd element of the rotated vector,
+                # regardless of orientation. For simplicity, we multiply by -1 so that acceleration
+                # during motor burn is positive, and acceleration due to drag force during coast
+                # phase is negative.
+                rotated_accelerations[idx] = -rotated_accel[2]
+
+            # Update the class attribute with the latest quaternion orientation
+            self._current_orientation_quaternions = current_orientation
+
+            return rotated_accelerations
+
+        # the quaternions are not uncertain, so we can directly use the quaternions given from
+        # the IMU.
+        current_orientation = np.array(
+            [
+                last_packet.estOrientQuaternionW,
+                last_packet.estOrientQuaternionX,
+                last_packet.estOrientQuaternionY,
+                last_packet.estOrientQuaternionZ,
+            ]
+        )
+        # rotate each compensated acceleration by the most recent quaternion orientation. This
+        # introduces some slight error, because the orientation for the last packet is not going
+        # to be the orientation for each packet, but it should be a negligible difference.
```

Comment:
after #99 the error would be even lower, but I would be interested in what the overall difference in predicted apogee is, when running the sim with `-f`.